presto-plan-checker-providers/src/main/java/com/facebook/presto/plancheckerproviders/nativechecker/NativePlanChecker.java

```diff
@@ -0,0 +1,250 @@
+        }
+        finally {
+            LOG.debug("Native plan validation complete for plan fragment id: %s", planFragment.getId());
+        }
+    }
+
+    private Request buildRequest(String requestBodyJson)
+    {
+        Request.Builder builder = new Request.Builder()
+                .url(getPlanValidateUrl())
+                .addHeader("CONTENT_TYPE", "APPLICATION_JSON")
+                .post(RequestBody.create(JSON_CONTENT_TYPE, requestBodyJson));
+
+        return builder.build();
+    }
+
+    private PlanConversionFailure processResponseFailure(Response response) throws IOException
+    {
+        if (response.body() != null) {
+            PlanConversionResponse planConversionResponse = PLAN_CONVERSION_RESPONSE_JSON_CODEC.fromJson(response.body().bytes());
+            if (!planConversionResponse.getFailures().isEmpty()) {
+                return planConversionResponse.getFailures().get(0);
+            }
+            else {
+                throw new PrestoException(GENERIC_INTERNAL_ERROR, "Error response without failure from native plan checker with code: " + response.code());
+            }
+        }
+        else {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "Error response without body from native plan checker with code: " + response.code());
+        }
+    }
+
+    private static class CheckInternalVisitor
+            extends PlanVisitor<Boolean, Void>
+    {
+        @Override
+        public Boolean visitTableScan(TableScanNode tableScan, Void context)
+        {
+            TableHandle handle = tableScan.getTable();
+            return ConnectorId.isInternalSystemConnector(handle.getConnectorId());
+        }
+
+        @Override
+        public Boolean visitPlan(PlanNode node, Void context)
+        {
+            for (PlanNode child : node.getSources()) {
+                if (child.accept(this, context)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+    }
+
+    public static class PlanConversionResponse
+    {
+        private final String status;
+        private final List<PlanConversionFailure> failures;
+
+        @JsonCreator
+        public PlanConversionResponse(
+                @JsonProperty("status") String status,
+                @JsonProperty("failures") List<PlanConversionFailure> failures)
+        {
+            this.status = requireNonNull(status, "status is null");
+            this.failures = ImmutableList.copyOf(requireNonNull(failures, "failures is null"));
+        }
+
+        @JsonProperty
+        public String getStatus()
+        {
+            return status;
+        }
+
+        @JsonProperty
+        public List<PlanConversionFailure> getFailures()
+        {
+            return failures;
+        }
+    }
+
+    public static class PlanConversionFailure
+            implements ErrorCodeSupplier
```

Comment:
Implementing `ErrorCodeSupplier` allows this to use the error code in the PrestoException